### PR TITLE
Security Guide

### PR DESCRIPTION
**Pull request by Mr0grog**
**WORK-IN-PROGRESS—not ready to merge**

This is a new guide to basic security practices with Stellar. At the moment, it’s mainly a *very* rough draft with lots of holes that was originally written over on Google Docs. All further work and discussion will be here.

Things that need to be addressed here:

- [x] Add section on revocable flag, per @bartekn (you should review this, Bartek)
- [x] Open question in compromised keys section about accounts on which the compromised account is a signer
- [x] Bugs in Stellar code—is this good or too thin?
- [x] Securing Core—this section is barely even a rough draft. It needs more details to be filled in (@jedmccaleb, @bartekn, @nullstyle)
- [x] Keeping up to date—I just put down some ideas here, but have no idea what the actual story is supposed to be. This needs correct info. (@jedmccaleb, @bartekn, @nullstyle)
- [x] Individuals—does this section even belong here? What should be in it?
- [x] Compliance
- [x] Add info about Stellar’s built-in security (ed25519, etc.)

There’s also sort of an open question as to organization. Right now this is a flat list, but could be organized into sections, maybe like:

- Accounts
  - Secure offline accounts
  - Require multiple authorizations or signers
  - Ensure assets are revocable
- Core
  - Securing a Stellar Core Instance
  - Keep Up to Date with Security Patches
- Handling Problems
  - What if an account’s keys are compromised?
  - What if there’s a bug in Stellar’s code?

Is that a better way to organize this doc?

### CODE REVIEW

File: guides/security.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+## What if an account’s keys are compromised?
+
+Because Stellar’s security is based around public key encryption, it’s critical that an account’s secret seed is not shared. Anyone who has access to the seed effectively has control of the account. However, if someone learns your account’s seed or you accidentally share it with someone who shouldn’t know it, you can remove its ability to control the account with the following steps:
+
+1. Make a new account.
+2. Add the new account as a signer on the compromised account. (Use the [`set options` operation](concepts/list-of-operations.md#set-options).
+3. Remove the compromised key’s signing authority on the compromised account.
+4. Now the new account controls the compromised account and the compromised account’s keys are no longer able to sign transactions.
+
+(NOTE: You must notify the owners of other accounts that the compromised account has signing authority or that the keys are compromised. They need to remove the compromised account as a signer as well [and add the new account].)
```
Comment (Mr0grog) [inline guides/security.md]: (Discussion from Google Docs…)

@mr0grog: Is this accurate? Or is signing transitive in this way—if account A requires signing from account B (the compromised account), which cannot sign for itself but instead requires signing from account C, does A require signing from C?

(This is prompted by the third paragraph here, which is not quite clear to me in this regard: https://www.stellar.org/developers/learn/concepts/multi-sig.html#transaction-signatures)

@bartekn: Not sure if I understand the question correctly so if it doesn't make sense let me know:

There are multiple procedures on how signatures can be added:

1. Signers can sign a transaction and send envelope to the "next" signer, after adding the last required signature tx can be submitted to the network.
2. Each signer can add a signature and send an envelope to a "central server" that will merge all envelopes and generate a single envelope with all signatures.

Answering your question:
For the method 1) signers should be informed about invalidated keys (or simply should check the current list of signers using Horizon) because some malicious user can sign any transaction and, by sending it to the next signer, trigger the chain of signatures.

For the method 2) there is no need to inform every signer because the final envelope will be created by a single user who should know which signers are valid.

@mr0grog: Hmmm, so I think what I was getting at was a scenario like this:

1. Account A requires signatures from accounts B and C to perform a transaction.

2. C's keys are compromised, so C goes through the procedure listed here to create account D, which will have sole signing authority over C (so the old C keys no longer work).

3. At this point, what happens for A? Do the old C keys work to sign a transaction by A? (I would think not, but want to make sure.) Can D sign a transaction from A by virtue of being the only signer on C? Or does A need to remove C as a signer and add D?

In other words, is it **safe** for A to do nothing? (So long as the original owner of C was responding ASAP, of course.) Or does A need to make sure to respond as well by removing C? (They probably should, but the question is whether they have to.)

File: guides/security.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+Signing of releases from Stellar.org?
+Any kind of feed for releases?
+Watch the forums?
+????
+
+
+## For Individuals
+
+TODO: Use a wallet? Encrypt your secret seed with a password? What else?
```
Comment (Mr0grog) [inline guides/security.md]: Does this section belong at all?

File: guides/security.md (hunk 1 above)
Comment (bartekn) [inline guides/security.md]: Sorry, I missed your response.

> At this point, what happens for A? Do the old C keys work to sign a transaction by A? (I would think not, but want to make sure.) Can D sign a transaction from A by virtue of being the only signer on C? Or does A need to remove C as a signer and add D?

It all depends what's the state of the account in the ledger. If the compromised keys are already removed from signers then, no matter what A does, network will simply reject the transaction.

Comment (Mr0grog) [inline guides/security.md]: @bartekn:

> It all depends what's the state of the account in the ledger. If the compromised keys are already removed from signers

Yeah, I guess I’m trying to understand the optimistic case here, where the owner of C has already removed the original key as a signer. Obviously if that hasn’t happened, A is vulnerable to malicious use by a third party who has obtained the original keys to C.

> no matter what A does, network will simply reject the transaction.

Is that because I said “A requires signatures from accounts B **and** C?” So, to make sure I’m getting this, D *cannot* sign a transaction for A on behalf of C? To put it another way: does the list of signers for A denote *public keys* that can sign and NOT *accounts* that can sign? Or is it something more complicated than that?

In my head I was actually thinking “B **or** C” (I mistyped, sorry). Does that change things?

File: guides/security.md (hunk 2 above)
Comment (jedmccaleb) [inline guides/security.md]: I don't think so.

File: guides/security.md [hunk 3]
```diff
@@ -30,12 +30,12 @@ If you issue your own assets, you should usually ensure that they can be revoked
 
 Because Stellar’s security is based around public key encryption, it’s critical that an account’s secret seed is not shared. Anyone who has access to the seed effectively has control of the account. However, if someone learns your account’s seed or you accidentally share it with someone who shouldn’t know it, you can remove its ability to control the account with the following steps:
 
-1. Make a new account.
-2. Add the new account as a signer on the compromised account. (Use the [`set options` operation](concepts/list-of-operations.md#set-options).
+1. Make a new key pair.
+2. Add the new public key as a signer on the compromised account. (Use the [`set options` operation](concepts/list-of-operations.md#set-options).
```
Comment (Mr0grog) [inline guides/security.md]: Oh! I was totally under the impression from an earlier conversation w/ @bartekn that all the signers had to be references to existing accounts, not just keys! That definitely makes this process seem a little less unwieldy.

This also clarifies the confusion I was having here: https://github.com/stellar/docs/pull/149/#discussion_r70694545 — if the signers are just keys, all accounts that accept a compromised key as a signer *do* need to remove it, since there’s no possibility for a transitive signing relationship if the signers are just keys and not references to accounts.

File: guides/security.md (hunk 1 above)
Comment (Mr0grog) [inline guides/security.md]: Ok, @jedmccaleb changes in ca1a1eccb5acd203b41791676597e917e3c0e06f clarified this for me. I had thought the signers were references to accounts, but they’re just public keys, so there’s no way Stellar could figure out that D could sign on behalf of C.

**Comment by Mr0grog** [comment]
Ok, so there are three things left here:

- Is “What if there’s a bug in Stellar’s code” section good enough?
- Need details for Stellar Core section
- Need *something* concrete to say for the “keeping up to date” section. If it’s just “keep an eye on Slack,” that’s fine. Is that the only story we have for updates?

**Comment by jedmccaleb** [comment]
> Is “What if there’s a bug in Stellar’s code” section good enough?
Yeah seems good.

> Need details for Stellar Core section
Yeah it doesn't need to be that secure really but I'll try to add some details.

> Need something concrete to say for the “keeping up to date” section.
We have a mailing list. They should subscribe to sdf-releases@freelists.org https://www.freelists.org/list/sdf-releases

**Comment by Mr0grog** [comment]
@jedmccaleb I cleared out the unfinished bits of section on core. You should give this a once-over to make sure it sounds good, then it ought to be ready to :shipit:

**Comment by bartekn** [comment]
:+1:
